### PR TITLE
Alchemist driver support

### PR DESCRIPTION
**Pull request by jstefanop**

### CODE REVIEW

File: README.ASIC [hunk 1]
```diff
@@ -6,6 +6,24 @@ Butterfly Labs' SC range of devices, HashBuster boards, GekkoScience's Compac
 USB stick, Klondike modules, and KnCMiner's Mercury, Jupiter and Saturn.
 
 
+ALCHEMIST
+---------
+
+This driver has been designed to run each of the 8 blades inside an AlcheMist 256 as a separate miner. To detect all the blades you need to manually probe it with the following serial ports: 
+
+-S ALC:all -S ALC:/dev/ttyO1 -S ALC:/dev/ttyO2 -S ALC:/dev/ttyO3 -S ALC:/dev/ttyO4
```
Comment (luke-jr) [inline README.ASIC:13]: Any reason not to add ttyO\* to autoscan?

File: README.ASIC [hunk 2]
```diff
@@ -6,6 +6,24 @@ Butterfly Labs' SC range of devices, HashBuster boards, GekkoScience's Compac
...
+
+(the four ttyUSB ports are auto detected by all)
+
+The driver supports custom frequency settings in the range of 200-400mhz in 16mhz increments(driver will default to 352mhz if these conditions are not met). Frequency is set with the following —-set option:
+
+--set ALC:clock=336
+
+You can also set the frequency per board by specifying the tty port: --set ALC@/dev/ttyO3:clock=352
+
+Driver also supports running single blades off a raspberry pi. Make sure you have configured gpio pin 25 correctly and scan via -S ALC:/dev/ttyAMA0
```
Comment (luke-jr) [inline README.ASIC]: What does it mean to have gpio pin 25 configured correctly?

File: configure.ac [hunk 3]
```diff
@@ -553,6 +553,15 @@ else
 	adl="no"
 fi
 
+BFG_DRIVER(,AlcheMist,scrypt,auto,[
+	need_lowl_vcom=yes
+	need_lowl_ftdi=yes
```
Comment (luke-jr) [inline configure.ac]: This is only for lowl-ftdi / ft232\* functions. You don't seem to use those.

File: configure.ac [hunk 4]
```diff
@@ -553,6 +553,15 @@ else
...
+	need_lowl_usb=yes
```
Comment (luke-jr) [inline configure.ac]: I don't think you need lowl-usb either?

File: configure.ac [hunk 5]
```diff
@@ -553,6 +553,15 @@ else
...
+	has_asic=yes
+	has_fpga=yes
```
Comment (luke-jr) [inline configure.ac]: FPGA?

File: configure.ac [hunk 6]
```diff
@@ -553,6 +553,15 @@ else
...
+	$broad_udevrules && have_udevrules=true
```
Comment (luke-jr) [inline configure.ac]: You're not adding any udev rules to 70-bfgminer.rules.in?

File: driver-alchemist.c [hunk 7]
```diff
@@ -0,0 +1,561 @@
+ /*
+ * Copyright 2015 jstefanop
```
Comment (luke-jr) [inline driver-alchemist.c]: Please also add yourself to AUTHORS. (Maybe use your full name? Up to you...)

File: driver-alchemist.c [hunk 8]
```diff
@@ -0,0 +1,561 @@
...
+ * Copyright 2014-2015 Luke Dashjr
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 3 of the License, or (at your option)
+ * any later version.  See COPYING for more details.
+ */
+
+//THIS DRIVER REQUIRES THE LATEST FIRMWARE FOR AlCHEMINERS DEVELOPED BY JSTEFANOP
+//IT WILL NOT WORK WITH THE STOCK FACTORY FIRMWARE ON THE BOARDS
+//PLEASE CONTACT JSTEFANOP AT MAC DOT COM OR JSTEFANOP ON LITECOINTALK DOT ORG FOR MORE INFO
```
Comment (luke-jr) [inline driver-alchemist.c:12]: This should probably be moved to README.ASIC?

File: driver-alchemist.c [hunk 9]
```diff
@@ -0,0 +1,561 @@
...
+};
+
+static
+void alchemist_chip_init(struct alchemist_chip * const chip, const uint8_t chipid)
+{
+	*chip = (struct alchemist_chip){
+		.chipid = chipid,
+		.global_reg = {0, 4, 0x40, 0, 0, 0, 0, 1},
+        .chip_mask = {0x0000, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000},
+        .clst_offset = {0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000, 0x00000000},
```
Comment (luke-jr) [inline driver-alchemist.c:69]: Unspecified fields get initialised to zero. Any reason to be explicit here?

File: driver-alchemist.c [hunk 10]
```diff
@@ -0,0 +1,561 @@
...
+    if(strcmp(devpath, "/dev/ttyUSB0") == 0)
+        strcpy(gpio, "gpio113");
+    if(strcmp(devpath, "/dev/ttyUSB1") == 0)
+        strcpy(gpio, "gpio114");
+    if(strcmp(devpath, "/dev/ttyUSB2") == 0)
+        strcpy(gpio, "gpio115");
+    if(strcmp(devpath, "/dev/ttyUSB3") == 0)
+        strcpy(gpio, "gpio116");
+    if(strcmp(devpath, "/dev/ttyAMA0") == 0)
+        strcpy(gpio, "gpio25");
```
Comment (luke-jr) [inline driver-alchemist.c:100]: If nothing matches, gpio is used uninitialised!

File: driver-alchemist.c [hunk 11]
```diff
@@ -0,0 +1,561 @@
...
+    unsigned char swpdata[80];
+    
+    buf[0] = 0xff;
+    
+    memset(&buf[1], 0, 0x18);
+    memcpy(&buf[25], &target[24], 0x8);
+    //pk_u64be(buf, 25, 0x0000feff01000000);
+    
+    swap32tobe(swpdata, work->data, 80/4);
+    memcpy(&buf[33], swpdata, 80);
```
Comment (luke-jr) [inline driver-alchemist.c:241]: Why not swap directly to buf?

**Comment by luke-jr** [comment]
A lot of mixed indentation in the new driver code; can you change it to be consistently tabs easily?

**Comment by luke-jr** [comment]
(Also, please address the Travis failures.)

File: driver-alchemist.c (hunk 8 above)
Comment (jstefanop) [inline driver-alchemist.c:12]: done

File: driver-alchemist.c (hunk 9 above)
Comment (jstefanop) [inline driver-alchemist.c:69]: to keep my sanity when i was debugging ;)

File: driver-alchemist.c (hunk 11 above)
Comment (jstefanop) [inline driver-alchemist.c:241]: needed to debug data block

File: driver-alchemist.c (hunk 7 above)
Comment (jstefanop) [inline driver-alchemist.c]: done

File: configure.ac (hunk 6 above)
Comment (jstefanop) [inline configure.ac]: removed all non needed config options

**Comment by jstefanop** [comment]
not sure whats causing indentation issues...looks fine on my system. Fixed travis issues with my code...not sure why its failing now seems like some include problems?

**Comment by luke-jr** [comment]
config.h must always be included first.

**Comment by luke-jr** [comment]
Merged as 67312a55090f2740310ce0c7c2bdfe5505f231cc

**Pull request closed without merging**